### ISSUE TITLE
Logs is not accessible due to missing log4j.properties and missing logback-console.xml in the auto generated flink-config-<cluster-name> configmap using the operator

### ISSUE THREAD

**Issue by FuyaoLi2017** [wangyang0918/flink-native-k8s-operator#4]
When I run the operator using both Flink CLI and the operator, it will give me such sed error at the beginning of the logs.

When I use Flink CLI based way to start a flink application. logs is still available. However, when I use this operator, the logs is not available.

When I create the flink application using FLINK CLI. The flink-config-volume will contain three files
`flink-conf.yaml` , `log4j-console.properties` and `logback-console.xml` .

However, when I tried to create flink application using this operator. With the same image, the auto generated flink configmap only contains `flink-conf.yaml` and there is NO `log4j-console.properties` and `logback-console.xml`

Here are the arguments in JM pod. As you can see, the operator created pod's command is missing log related configurations.

Flink CLI:
```
Args:
      native-k8s
      $JAVA_HOME/bin/java -classpath $FLINK_CLASSPATH -Xmx1073741824 -Xms1073741824 -XX:MaxMetaspaceSize=268435456 -Dlog.file=/opt/flink/log/jobmanager.log -Dlogback.configurationFile=file:/opt/flink/conf/logback-console.xml -Dlog4j.configuration=file:/opt/flink/conf/log4j-console.properties -Dlog4j.configurationFile=file:/opt/flink/conf/log4j-console.properties org.apache.flink.kubernetes.entrypoint.KubernetesApplicationClusterEntrypoint -D jobmanager.memory.off-heap.size=134217728b -D jobmanager.memory.jvm-overhead.min=201326592b -D jobmanager.memory.jvm-metaspace.size=268435456b -D jobmanager.memory.heap.size=1073741824b -D jobmanager.memory.jvm-overhead.max=201326592b
```

Operator created JM:

```
    Args:
      native-k8s
      $JAVA_HOME/bin/java -classpath $FLINK_CLASSPATH -Xmx3462817376 -Xms3462817376 -XX:MaxMetaspaceSize=268435456 org.apache.flink.kubernetes.entrypoint.KubernetesApplicationClusterEntrypoint -D jobmanager.memory.off-heap.size=134217728b -D jobmanager.memory.jvm-overhead.min=429496736b -D jobmanager.memory.jvm-metaspace.size=268435456b -D jobmanager.memory.heap.size=3462817376b -D jobmanager.memory.jvm-overhead.max=429496736b

```

I do keep the command template in the flinkConfig part in the `cr.yaml` file. It doesn't help due to some unknown reasons.

```
kubernetes.container-start-command-template: "%java% %classpath% %jvmmem% %jvmopts% %logging% %class% %args%"
```
Please share some ideas.. Thanks!

I also checked this thread. Didn't find anything useful. Any ideas on this? Thanks.
http://apache-flink.147419.n8.nabble.com/flink-1-11-on-kubernetes-td4586.html

```bash

$ kubectl logs -f flink-demo-56f487f446-5w86h
sed: couldn't open temporary file /opt/flink/conf/sedx3urji: Read-only file system
sed: couldn't open temporary file /opt/flink/conf/seduhOSTg: Read-only file system
/docker-entrypoint.sh: line 75: /opt/flink/conf/flink-conf.yaml: Read-only file system
sed: couldn't open temporary file /opt/flink/conf/sedrn0ZFj: Read-only file system
/docker-entrypoint.sh: line 88: /opt/flink/conf/flink-conf.yaml: Read-only file system
/docker-entrypoint.sh: line 90: /opt/flink/conf/flink-conf.yaml.tmp: Read-only file system
Start command: $JAVA_HOME/bin/java -classpath $FLINK_CLASSPATH -Xmx3462817376 -Xms3462817376 -XX:MaxMetaspaceSize=268435456 org.apache.flink.kubernetes.entrypoint.KubernetesApplicationClusterEntrypoint -D jobmanager.memory.off-heap.size=134217728b -D jobmanager.memory.jvm-overhead.min=429496736b -D jobmanager.memory.jvm-metaspace.size=268435456b -D jobmanager.memory.heap.size=3462817376b -D jobmanager.memory.jvm-overhead.max=429496736b
ERROR StatusLogger No Log4j 2 configuration file found. Using default configuration (logging only errors to the console), or user programmatically provided configurations. Set system property 'log4j2.debug' to show Log4j 2 internal initialization logging. See https://logging.apache.org/log4j/2.x/manual/configuration.html for instructions on how to configure Log4j 2
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.apache.flink.api.java.ClosureCleaner (file:/opt/flink/lib/flink-dist_2.11-1.12.1.jar) to field java.util.Properties.serialVersionUID
WARNING: Please consider reporting this to the maintainers of org.apache.flink.api.java.ClosureCleaner
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

**Comment by FuyaoLi2017** [comment]
I think the flink job is indeed running. It just doesn't print out logs with the `kubectl logs -f <pod>` command. Maybe I should create some sort of sidecar container to make it work?

**Comment by wangyang0918** [comment]
I remember from 1.12, native Flink K8s cluster will output the logs to local files and console simultaneously. For 1.11, you need to update your local log4j.properties.

https://ci.apache.org/projects/flink/flink-docs-release-1.11/ops/deployment/native_kubernetes.html#log-files

**Comment by FuyaoLi2017** [comment]
Hello Yang, could you check the flink mailing list email with name `StopWithSavepoint() method doesn't work in Java based flink native k8s operator` It is describing the same issue and I attached some reference information. I also cc you in the thread.

My modified implementation:  https://github.com/FuyaoLi2017/flink-native-kubernetes-operator/blob/master/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkApplicationController.java

I think there is an issue with the operator itself.

When I create the flink application using FLINK CLI. The flink-config-volume will contain three files 
`flink-conf.yaml` , `log4j-console.properties` and `logback-console.xml` .

However, when I tried to create flink application using this operator. With the same image, the auto generated flink configmap only contains `flink-conf.yaml` and there is NO `log4j-console.properties` and `logback-console.xml` when I exec into the pod... Very confused.


Also, could you please check why stopWithSavepoint() issue mentioned in the thread. I can only terminate the native k8s application to do image update using cancelWithSavepoint(). If I use stopWithSavepoint(), it will hang there forever and never complete the CompleteableFuture. TriggerSavepoint() and cancelWithSavepoint() could work well.

https://github.com/FuyaoLi2017/flink-native-kubernetes-operator/blob/master/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkApplicationController.java#L262

Looking forward to your reply. Thank you sooo much!

**Comment by FuyaoLi2017** [comment]
If possible, could you add my wechat and we can have a short discussion. wechat: `1605419584` . If you feel inconvenient, it is totally fine. Thanks!

**Comment by FuyaoLi2017** [comment]
I am expecting this class could automatically mount the logs related configmaps.
https://github.com/apache/flink/blob/0350bf105c41e92599fc64952610187d2385eb28/flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/decorators/FlinkConfMountDecorator.java

However, it seems that it only mounts the flink-conf.yaml and doesn't take the other two into account.

**Comment by wangyang0918** [comment]
@FuyaoLi2017 Do you mean `FlinkConfMountDecorator#getLocalLogConfFiles` does not find the `log4j-console.properties` and `logback-console.xml`. The log configuration files should be located under `/opt/flink/conf`. Right?

**Comment by wangyang0918** [comment]
For `ClusterClient#stopWithSavepoint`, could you please check that whether you could use the Flink CLI to stop the application? And the JobManager logs will help a lot for debugging this issue.

**Comment by FuyaoLi2017** [comment]
> For `ClusterClient#stopWithSavepoint`, could you please check that whether you could use the Flink CLI to stop the application? And the JobManager logs will help a lot for debugging this issue.

For the ClusterIP type of deployment, as we have discussed in another thread, we can't directly use Flink CLI. We should use flink client deployment to execute the Flink CLI (kubectl exec -it ...). Maybe due to some RABC related issues, I can't manage the operator-created Flink application with Flink CLI in the Flink client pod.

For the pure Flink native k8s deployment with LoadBalancer Mode. I can do list command, but not stop/cancel/savepoint command. I will check the logs... The interesting thing is that I can trigger the savepoint via Flink REST API directly in the JM pod. However, even the savepoint is triggered, the job and pods are not terminated, but no new checkpoints is generated. Some details can be found the email thread in the flink user mailing list. I have also cc you there. Thanks!

**Comment by FuyaoLi2017** [comment]
Hello Yang, I figured out why I can't execute 
- stop/cancel/savepoint command in Flink CLI 
- why I can't execute the stopWithSavepoint.

For the question 1,
I was directly following the commands in https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/deployment/cli/#cancelling-a-job-ungracefully
without adding `--target kubernetes-application -Dkubernetes.cluster-id=<cluster id>`.

After adding it back. It can execute all Flink CLI commands.

For question 2, 
Since I can't view the logs directly because of the log issue I mentioned. I tried to deploy the same image and trigger stopWithSavepoint there after understanding question 1. I found that due to the Kafka version is too low. I used 1.1.0 as my Kafka-client in the flink application. 
As a result, I noticed such error.
```
java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: 'void org.apache.kafka.clients.producer.KafkaProducer.close(java.time.Duration)'
```

I fixed it after upgrading the Kafka client version.

For the log issue, I still can't figure it out by myself. Any suggestions?

**Comment by FuyaoLi2017** [comment]
> @FuyaoLi2017 Could you please share the JobManager pod start command? I think you could use `kubect get pods {jm_pod_name} -o yaml` to get the command.
> After then, we could check whether the log4j properties is correctly set.

Hello Yang, I have update the question description last weekend. Please refer to the question description for the full command.

for the operator created flink application. Full Yaml is shared at the bottom of the message.

Those flink application created by operator (with the same image) is missing all logs related configuration shared below.
```
-Dlog.file=/opt/flink/log/jobmanager.log -Dlogback.configurationFile=file:/opt/flink/conf/logback-console.xml -Dlog4j.configuration=file:/opt/flink/conf/log4j-console.properties -Dlog4j.configurationFile=file:/opt/flink/conf/log4j-console.properties
```

The configmap for Flink CLI created application will have three entry (flink-conf.yaml, ) for the flink CLI automatically created configmap flink-config-<cluster-id>.

There is only 1 for the operator automatically created configmap (only flink-conf.yaml)

```
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2021-05-11T01:50:11Z"
  generateName: flink-demo-664648bd45-
  labels:
    app: flink-demo
    component: jobmanager
    pod-template-hash: 664648bd45
    type: flink-native-kubernetes
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:generateName: {}
        f:labels:
          .: {}
          f:app: {}
          f:component: {}
          f:pod-template-hash: {}
          f:type: {}
        f:ownerReferences:
          .: {}
          k:{"uid":"a3f0f972-0f93-484e-b607-55c281ee101c"}:
            .: {}
            f:apiVersion: {}
            f:blockOwnerDeletion: {}
            f:controller: {}
            f:kind: {}
            f:name: {}
            f:uid: {}
      f:spec:
        f:containers:
          k:{"name":"flink-job-manager"}:
            .: {}
            f:args: {}
            f:command: {}
            f:env:
              .: {}
              k:{"name":"_POD_IP_ADDRESS"}:
                .: {}
                f:name: {}
                f:valueFrom:
                  .: {}
                  f:fieldRef:
                    .: {}
                    f:apiVersion: {}
                    f:fieldPath: {}
            f:image: {}
            f:imagePullPolicy: {}
            f:name: {}
            f:ports:
              .: {}
              k:{"containerPort":6123,"protocol":"TCP"}:
                .: {}
                f:containerPort: {}
                f:name: {}
                f:protocol: {}
              k:{"containerPort":6124,"protocol":"TCP"}:
                .: {}
                f:containerPort: {}
                f:name: {}
                f:protocol: {}
              k:{"containerPort":8081,"protocol":"TCP"}:
                .: {}
                f:containerPort: {}
                f:name: {}
                f:protocol: {}
            f:resources:
              .: {}
              f:limits:
                .: {}
                f:cpu: {}
                f:memory: {}
              f:requests:
                .: {}
                f:cpu: {}
                f:memory: {}
            f:terminationMessagePath: {}
            f:terminationMessagePolicy: {}
            f:volumeMounts:
              .: {}
              k:{"mountPath":"/opt/flink/conf"}:
                .: {}
                f:mountPath: {}
                f:name: {}
        f:dnsPolicy: {}
        f:enableServiceLinks: {}
        f:imagePullSecrets:
          .: {}
          k:{"name":"ocirsecret"}:
            .: {}
            f:name: {}
        f:restartPolicy: {}
        f:schedulerName: {}
        f:securityContext: {}
        f:serviceAccount: {}
        f:serviceAccountName: {}
        f:terminationGracePeriodSeconds: {}
        f:volumes:
          .: {}
          k:{"name":"flink-config-volume"}:
            .: {}
            f:configMap:
              .: {}
              f:defaultMode: {}
              f:items: {}
              f:name: {}
            f:name: {}
    manager: kube-controller-manager
    operation: Update
    time: "2021-05-11T01:50:11Z"
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:conditions:
          k:{"type":"ContainersReady"}:
            .: {}
            f:lastProbeTime: {}
            f:lastTransitionTime: {}
            f:status: {}
            f:type: {}
          k:{"type":"Initialized"}:
            .: {}
            f:lastProbeTime: {}
            f:lastTransitionTime: {}
            f:status: {}
            f:type: {}
          k:{"type":"Ready"}:
            .: {}
            f:lastProbeTime: {}
            f:lastTransitionTime: {}
            f:status: {}
            f:type: {}
        f:containerStatuses: {}
        f:hostIP: {}
        f:phase: {}
        f:podIP: {}
        f:podIPs:
          .: {}
          k:{"ip":"10.244.1.14"}:
            .: {}
            f:ip: {}
        f:startTime: {}
    manager: kubelet
    operation: Update
    time: "2021-05-11T01:50:13Z"
  name: flink-demo-664648bd45-2v7rg
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: flink-demo-664648bd45
    uid: a3f0f972-0f93-484e-b607-55c281ee101c
  resourceVersion: "17732857"
  selfLink: /api/v1/namespaces/default/pods/flink-demo-664648bd45-2v7rg
  uid: 540d61c5-9745-4c8b-aba1-45c19c134f4e
spec:
  containers:
  - args:
    - native-k8s
    - $JAVA_HOME/bin/java -classpath $FLINK_CLASSPATH -Xmx3462817376 -Xms3462817376
      -XX:MaxMetaspaceSize=268435456 org.apache.flink.kubernetes.entrypoint.KubernetesApplicationClusterEntrypoint
      -D jobmanager.memory.off-heap.size=134217728b -D jobmanager.memory.jvm-overhead.min=429496736b
      -D jobmanager.memory.jvm-metaspace.size=268435456b -D jobmanager.memory.heap.size=3462817376b
      -D jobmanager.memory.jvm-overhead.max=429496736b
    command:
    - /docker-entrypoint.sh
    env:
    - name: _POD_IP_ADDRESS
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    image: us-phoenix-1.ocir.io/idxglh0bz964/flink-demo:new-kafka-version-2
    imagePullPolicy: IfNotPresent
    name: flink-job-manager
    ports:
    - containerPort: 8081
      name: rest
      protocol: TCP
    - containerPort: 6123
      name: jobmanager-rpc
      protocol: TCP
    - containerPort: 6124
      name: blobserver
      protocol: TCP
    resources:
      limits:
        cpu: "1"
        memory: 4Gi
      requests:
        cpu: "1"
        memory: 4Gi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /opt/flink/conf
      name: flink-config-volume
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: flink-native-k8s-operator-token-lvnmg
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  imagePullSecrets:
  - name: ocirsecret
  nodeName: 10.196.109.53
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: flink-native-k8s-operator
  serviceAccountName: flink-native-k8s-operator
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - configMap:
      defaultMode: 420
      items:
      - key: flink-conf.yaml
        path: flink-conf.yaml
      name: flink-config-flink-demo
    name: flink-config-volume
  - name: flink-native-k8s-operator-token-lvnmg
    secret:
      defaultMode: 420
      secretName: flink-native-k8s-operator-token-lvnmg
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-05-11T01:50:12Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2021-05-11T01:50:13Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2021-05-11T01:50:13Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2021-05-11T01:50:12Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://4292783bd2016939a3ca2075d80db9d72f07979fd3170bef6cfd4b13f0a69682
    image: us-phoenix-1.ocir.io/idxglh0bz964/flink-demo:new-kafka-version
    imageID: docker-pullable://us-phoenix-1.ocir.io/idxglh0bz964/flink-demo@sha256:775b71749024e4e096cdfdb650469e55e17092c8b6f44ec2436536430429a1ab
    lastState: {}
    name: flink-job-manager
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2021-05-11T01:50:12Z"
  hostIP: 10.196.109.53
  phase: Running
  podIP: 10.244.1.14
  podIPs:
  - ip: 10.244.1.14
  qosClass: Guaranteed
  startTime: "2021-05-11T01:50:12Z"
```

**Comment by wangyang0918** [comment]
Fixed via #5.
